### DISCUSSION TITLE
Refactoring has started!

### DISCUSSION THREAD

**Discussion by lambtho12** [nvim-telekasten/telekasten.nvim#239] [General]
Hi all!

I finally managed to set aside a bit of time every week-end to work on Telekasten! Yay!

I plan to proceed as follows:
- [x] Improve configuration management to avoid passing all options everywhere (get [inspiration from bufferline](https://github.com/akinsho/bufferline.nvim/blob/main/lua/bufferline/config.lua))
- [ ] Splitting the gigantic main file into smaller chunks so it is easier to manage
- [ ] Use plenary.path wherever possible to avoid issues with path or filenames (and hopefully better accommodate Windows users)
- [ ] Review link handling, remove hard-coded stuff and add full markdown-style link support
- [ ] Simplify syntax highlighting and maybe deprecate specific filetype(?) 
- [ ] Refactor a lot of stuff to mostly get rid of the duplicate code
- [ ] Improve regex for robustness

Hopefully by working on all this, I should simultaneously fix a few bugs and issues here and there. I will try to progress slowly and with caution in order to avoid any unfortunate breakage on your side. Ideally it would be nice to have some testers that would install the `refactor` branch and validate the changes I make before I merge them.

So, to be clear, I do not plan on adding new features straight away. I think the first step is really to make the code more "maintainable" first so maybe we can get more people involved as well.

**Comment by lambtho12** [comment]
Starting refactoring the configuration management a bit (see https://github.com/renerocksai/telekasten.nvim/commit/683f51217b9a07f44ac3cf6d9f664b18b2b04a5f). Hopefully I can get this done by the end of the week-end.

The configuration itself will not change, so no breaking changes or anything like that are to be expected. I basically want to move it all out of the main file and make it easier to access from other modules. Now I have to edit the billion calls to `M.Cfg` scattered everywhere...

**Reply by lambtho12** [reply]
Almost done. The refactor branch should be stable. It would be great if someone had the time to test it a bit.

**Comment by TheTaoOfSu** [comment]
I had some delays due to the holidays and needing to spend some time learning the internals of NeoVim first, but I'm getting started on the refactor. I'm starting by trying to get main merged into refact. General process for the bit I've done so far has been to take new code from main but make sure it's using the names as changed in refact when, e.g., config was pushed into a separate file. So far, I haven't seen any new functionality in refact, which makes sense since it's a refactor branch, so there hasn't been any real issue with it, but I'm not super deep in.

Once I get that done, I want to go through and take a look at each function in telekasten.lua and summarize the key steps in what it's doing. With that, it should be easier to see how functions should be split, combined, and moved to new files. It will also be a good time to make sure paths are being handled by plenary.path(). Finally, after those, I should be able to handle the markdown links.

Syntax highlighting seems like it's probably going to require more specialized knowledge than I currently have or am likely to be able to develop soon, so I will likely need to leave that to someone else.

I'm hoping to finish merging main in the next couple days, but it may take until some time next week. Would really like to have the markdown links working ASAP, so I'm hoping I can keep chugging along at a good pace.

**Reply by TheTaoOfSu** [reply]
A lot happened in the last 3, almost 3.5, months, and very little of it has been conducive to coding. Sorry for falling off and not communicating, but it always seemed like I was just about to get my chance to get back to it. I restarted today, wiped the meager progress I'd made and restarted the process of merging main into refact. I'm about 15% through with sorting out the conflicts if we measure by line count, i.e., have cleared all conflicts in the first 15% of all lines in telekasten.lua, the only file to see conflicts. To recap, my process has essentially been to take all the functionality from main and adapt it to the new names and structuring of refact. If I keep going at this speed, I should hopefully have it done by tomorrow, maybe early next week if I don't finish and can't get much done over the weekend.

Once that's done, I want to go through and add some comments in prep for greater refactoring. Basically, look through each function, see what it's doing and how, and make sure there are good notes on that. Also note some of the stuff like what the unnamed functions are doing that get passed as "callback" to certain functions like global_dir_check and check_dir_and_ask so they can maybe be formally named and made to be proper functions. If anyone has any recommendations or requests for this phase, now would be an excellent time to voice them.

**Reply by lambtho12** [reply]
AWESOME!

I have very limited time available at the moment, but I will try my best to support you in this refactoring.

Functions such as `check_dir_and_ask` and `global_dir_check` are utilities that we use in order to check if the directories configured by the user are present and if not, ask the user if we should create them. These functions do not need to be exposed to the user. This is why I moved them in `telekasten/utils/files.lua`.

Regarding the refactoring, I think these are few good starting points:
- Improve the config workflow so we can properly separate the different aspects of the config (vaults, paths, fileoptions, themes, etc)
- Make the pickers more abstract so we can reuse the same basic picker instead of copy/pasting the same code in all functions
- A lot of small stuff is duplicated, transform it to atomic functions instead and reuse said function
- Move all utilities in proper files (e.g., `make_relative_path` in `utils/files.lua`). Eventually, the `telekasten.lua` file should only contain the user-facing functions.
- Improve all regexes so they are more robust
- Remove telekasten filetype (keep syntax). It causes more issues than what it was meant to fix.

**Reply by TheTaoOfSu** [reply]
My point about check_dir _and_ask and global_dir_check was more that they're often passed an unnamed, single use function as their callback parameter, and while I haven't gotten to examine them closely yet, these functions seem like they may be worth formalizing or whatever the proper word for it is. It at least seems worth evaluating the option if I'm refactoring anyway.

This list will make a good reference when deciding how to focus my efforts. A lot of this seems pretty doable, at least when I first hear about it. I'm not really sure about the regexes, though, so I may have to leave that for someone else, but I'll try to take a stab at it last if I still have time to spend on it. The filetype and syntax stuff is a little too vim-specific, and I don't see myself getting to it. I don't have the knowledge to do it, and I don't see myself having the time to devote to learning intricacies of neovim over other things I should be learning right now. If I complete the refactor and add the functionality I personally want and still have some time to burn on it, I'll give it a shot, but I consider it unlikely.

As for the config workflow, what are you thinking? My first thought is to split the config into different subtables to handle certain things, so like you'd have config.options.path that has paths for home, dailies, weeklies, etc.. I'd have to look over all the options in detail to make a real suggestion, but I feel like at least a good chunk of the options could be organized in subtables.

**Reply by lambtho12** [reply]
> these functions seem like they may be worth formalizing

True

> I'm not really sure about the regexes

Basically the idea would be to use stuff like `\p{L}` which are not language specific, rather and `[a-zA-Z]` for instance. This will shorten the expression, make it easier to understand, and will work in any alphabet (including chinese, thai, russian, etc). 

> The filetype and syntax stuff is a little too vim-specific, and I don't see myself getting to it. 

Ok, no problem. This is not the most urgent one anyway.

> Config workflow, [...] split the config into different subtables

This is precisely it!

**Reply by TheTaoOfSu** [reply]
I'm at 50% by line count, so I suppose it's about time to start talking about how best to handle the pull requests. First off, I'm thinking I should make a request for the main merge as soon as that's done and get refact updated in the official repo. From there, probably basically one request per bullet point as you listed it. Maybe split the util function moves into one per util file, but probably only if the work ends up being substantial. I feel like it shouldn't be too rough, though. Should I bother making a separate pull request for all the prep comments I'm planning to make next?

**Reply by lambtho12** [reply]
> make a request for the main merge as soon as that's done and get refact updated in the official repo. 

Definitely. That would be a nice starting point

> one request per bullet point as you listed it. Maybe split the util function moves into one per util file, but probably only if the work ends up being substantial

Yes. I would rather have to review 3 smaller PRs rather than a gigantic one. So if you feel it could be splitted, do not hesitate. 

> Should I bother making a separate pull request for all the prep comments I'm planning to make next?

It is up to you. I have no issue with just having a look at that and maybe take the opportunity to discuss a bit implementation before you start working.

**Reply by TheTaoOfSu** [reply]
Sounds good. As for the comments pull request, I'll probably do it just in case anything happens again and I don't get to finish. I'm hoping I do, but by that point, it's a lot of effort to figure all that out, and it might as well get documented. A lot of the decisions for the rest of it will either be made there or heavily simplified and influenced by the info they have.

**Reply by TheTaoOfSu** [reply]
I've finished a first pass at the merge. I had to resolve a lot of conflicts in telekasten.lua, add the media_extensions config option to telekasten/config.lua, and update the telekasten/util/files.lua functions check_dir_and_ask and global_dir_check with changes that got incorrectly merged back into telekasten.lua. Any other changes were automatically resolved by git and are so far unchecked. I want to look over everything again with fresh eyes tomorrow (hopefully) to make sure I didn't make and miss any mistakes. If you want to preview the changes, I've pushed them to my fork, and you can see the commit with these changes [here](https://github.com/TheTaoOfSu/telekasten.nvim/commit/3f498cfee643034060ff25becc6f0cb3ebfb8d31). Otherwise, I'm hoping I can get a pull request made tomorrow. Should be a very low chance of anything being newly broken.

**Reply by TheTaoOfSu** [reply]
Pull request has officially been made for merging main into refact. As noted in the request, I don't have a full vault of my own to do much testing with. I made a few basic notes and tested some basic stuff, but I'm not well equipped for extensive testing.

**Comment by lambtho12** [comment]
@renerocksai, looks like @TheTaoOfSu is getting pretty involved in Telekasten and doing significant work here.

You discussed previously the possibility to move Telekasten to an organization rather than a personal repo. This may drive more contributors to the project and simplify the development. 
I think now would be a good time to do that. Or do you want to wait a bit more for the refactoring to be merged in main?

Could you try to find some time to put that in motion? I can be an admin for the organization alongside you if you are ok with that so I can manage everything and not bother you anymore :smile:

**Reply by lambtho12** [reply]
Obviously, if @TheTaoOfSu is interested, we could make them a developer in the organization so they can directly push to some feature branches etc if they are interested.

**Reply by TheTaoOfSu** [reply]
I like the idea of continuing to help, but I can't guarantee I'll be able to commit a lot afterwards. I suppose it would depend on a few things like the direction Telekasten takes, how fast things move, and how much free time I keep long term for coding. Development hasn't been particularly fast paced lately, so I could probably at least help occasionally.

**Reply by renerocksai** [reply]
It is done! Telekasten is now free! @lambtho12 , I have made you an owner of this org. Feel free to manage anything (I mean it) to your liking. I cannot express my gratitude in words; but I am very grateful! Thank you all (incl. @TheTaoOfSu) for keeping telekasten alive, for all the hard work, and improving the status quo!

**Reply by lambtho12** [reply]
Thank you for starting this awesome project! I hope we can continue improving it and make it a bit more simple to use/expand for new users.
Enormous thanks to @TheTaoOfSu for picking up the refactoring work as well!

**Comment by TheTaoOfSu** [comment]
Is there some decent documentation on Telescope and/or Plenary?  Figuring things out straight from the code gets cumbersome when the functions use other internal functions that may themselves sometimes use other internal functions... It doesn't seem super important that I understand exactly how some of these functions work, but it'd be nice to.

**Reply by lambtho12** [reply]
Telescope has a [wiki](https://github.com/nvim-telescope/telescope.nvim/wiki) and a [developer documentation](https://github.com/nvim-telescope/telescope.nvim/blob/master/developers.md). 

For plenary, I do not think there is any proper documentation. [Their README](https://github.com/nvim-lua/plenary.nvim) just contains the basic module names and a few indications. Instead of directly looking into the implementation of each function to discover what it does, you could maybe just have a look at the unit tests to have an idea of what to expect without needing to review the entire implementation.

**Comment by TheTaoOfSu** [comment]
I just finished a first pass at commenting all the functions in telekasten.lua. Primarily, this focused on comments at the start of function definitions to explain a bit and note a potential destination file, although I did sprinkle in a few TODOs and mid-function clarifications along the way, so it's not all focused on the refactor. Some will likely be worth keeping, some will be cleared as TODOs are completed and the refactor progresses, but some I'm not sure. One such thing I included was a note of the parameters' and return values' types in the form "param type, param type,... -> return type, return type,...". Although my Lua knowledge is limited, its flexibility suggests this may be a little unusual for Lua code, but I personally find you can get a _lot_ of basic questions about functions answered relatively quickly if you can just clearly tell the types of the inputs and outputs combined with a decent function name, even if the rest of the comments are lacking or nonexistent.

I did notice that a large majority of functions seemed fit for utils/files.lua, so I'm going to step away for a bit before I reevaluate some of it and see if I've been neglecting some other choices. Either way, soon after, I'll start moving functions around and doing my best to make sure functions are still addressed appropriate wherever they may be called from.

I don't think a lot of these comments need to be pushed upstream, so I won't make a request for what I have right now. If you want to check it out, you can see the new telekasten.lua [here](https://github.com/TheTaoOfSu/telekasten.nvim/blob/refact/lua/telekasten.lua) or the diff view [here](https://github.com/TheTaoOfSu/telekasten.nvim/commit/3348a6283cd3e1c95532db786a156e22faeac2bf). I'm hoping to get all the functions moved tonight now that I know what most do and where they should go, so don't worry about it if you don't get to look at it soon.

**Reply by TheTaoOfSu** [reply]
I got the initial movement done. Almost every function that isn't directly callable by the user has been moved out of telekasten.lua. I did change my initial assessment about where a lot of the functions should go, so I think this will be a better layout than my first thoughts, but I'd be happy to take corrections or suggestions on any of it. I expect some of it will change as I dig deeper into the refactor, but we'll see how things go.

This did introduce some circular requires as different modules became dependent on each other, which was a simple fix, but my test and work flow are clunky enough that it led to some rather minor commits, so I recommend squashing the PR to keep clutter out of the commit history.

Speaking of PRs, I did make one for this first draft. As I mentioned in the PR, I still don't have a great testing library, so while I can confirm that the plugin loads without error and some basic functionality seems unchanged, it wouldn't hurt to test things out a bit.

I do want to continue the refactor, but I think I need to address some features I need before I can start really using Telekasten, mainly just markdown link support. That's the sort of thing that's probably going to need to get merged into main, so it may be a good time to think about just what needs to be done before we can merge refact back into main. There's still plenty of refactoring work to continue afterwards, so it wouldn't be the end of the refact branch, but at this point, it seems best to use the refact branch as the base for it.

**Reply by lambtho12** [reply]
Awesome work! I will have a look at all that during the week-end. Thanks again for your involvement.

**Comment by sarahyack** [comment]
Hi! As I said I would, I took a look at the `refact` branch, and I figured I could at least update my own contribution to fit the new system, and I think I did pretty successfully and opened a pull request to merge with the refact branch, #369 . This PR pretty much eliminates most of the hard coded stuff in the plugin (with the exception of image filenames), and updates the various file system helpers to improve how they function and integrate with the periodic system. Other than that, I don't believe I got any other specific goals for the refactor done, and I'm wondering what you'd like me to prioritize next moving forward?

**Reply by lambtho12** [reply]
It has been a long time since I was active in developing this, you probably have a better understanding of the project than I do now :smile: 
I will review the PR and merge it, but feel free to tackle anything else you find useful.

If you want to get a more in depth understanding of all the features, maybe start by looking into the various issues. Feel free to ping me in discussions or issues (or even directly by mail) if you want to exchange a bit more before starting with the code.
